dali/operators/reader/loader/nemo_asr_loader.cc

```diff
@@ -0,0 +1,156 @@
+#include "dali/core/error_handling.h"
+#include "dali/operators/decoder/audio/generic_decoder.h"
+#include "dali/operators/decoder/audio/audio_decoder_impl.h"
+#include "dali/operators/reader/loader/nemo_asr_loader.h"
+#include "dali/operators/reader/loader/utils.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/pipeline/util/lookahead_parser.h"
+#include "dali/util/file.h"
+
+namespace dali {
+
+namespace detail {
+
+void ParseManifest(std::vector<NemoAsrEntry> &entries, const std::string &json) {
+  detail::LookaheadParser parser(const_cast<char*>(json.c_str()));
+
+  DALI_ENFORCE(parser.PeekType() == kArrayType);
+  parser.EnterArray();
+
+  while (parser.NextArrayValue()) {
+    if (parser.PeekType() != kObjectType)
+      continue;
+    parser.EnterObject();
+    NemoAsrEntry entry;
+    while (const char* key = parser.NextObjectKey()) {
+      if (0 == detail::safe_strcmp(key, "audio_filepath")) {
+        entry.audio_filepath = parser.GetString();
+      } else if (0 == detail::safe_strcmp(key, "duration")) {
+        entry.duration = parser.GetDouble();
+      } else if (0 == detail::safe_strcmp(key, "offset")) {
+        entry.offset = parser.GetDouble();
+      } else if (0 == detail::safe_strcmp(key, "text")) {
+        entry.text = parser.GetString();
+      } else {
+        parser.SkipValue();
+      }
+    }
+    entries.emplace_back(std::move(entry));
+  }
+}
+
+}  // namespace detail
+
+void NemoAsrLoader::PrepareMetadataImpl() {
+  std::ifstream fstream(manifest_filepath_);
+  DALI_ENFORCE(fstream,
+               make_string("Could not open NEMO ASR manifest file: \"", manifest_filepath_, "\""));
+  std::string json((std::istreambuf_iterator<char>(fstream)), std::istreambuf_iterator<char>());
+  detail::ParseManifest(entries_, json);
+
+  DALI_ENFORCE(Size() > 0, "No files found.");
+  if (shuffle_) {
+    // seeded with hardcoded value to get
+    // the same sequence on every shard
+    std::mt19937 g(kDaliDataloaderSeed);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+  Reset(true);
+}
+
+void NemoAsrLoader::Reset(bool wrap_to_shard) {
+  current_index_ = wrap_to_shard ? start_index(shard_id_, num_shards_, Size()) : 0;
+  current_epoch_++;
+
+  if (shuffle_after_epoch_) {
+    std::mt19937 g(kDaliDataloaderSeed + current_epoch_);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+}
+
+void NemoAsrLoader::PrepareEmpty(AsrSample &sample) {
+  PrepareEmptyTensor(sample.audio);
+}
+
+void NemoAsrLoader::ReadSample(AsrSample& sample) {
+  auto &entry = entries_[current_index_++];
+
+  // handle wrap-around
+  MoveToNextShard(current_index_);
+
+  // metadata info
+  DALIMeta meta;
+  meta.SetSourceInfo(entry.audio_filepath);
+  meta.SetSkipSample(false);
+  sample.audio.SetMeta(meta);
+
+  // Ignoring copy_read_data_, Sharing data is not supported with this loader
+  // TODO(janton): do not create a new decoder each time (?)
```

Comment:
It would be hard as when you create a decoder you parse the data and selects the right underlying implementation appropriate for the given audio format.